### PR TITLE
allow serde serialization and `TryFrom<f64>` with `no_std`

### PR DESCRIPTION
**Pull request by mversic**
* `serde` serialization can now be used with `no_std` (would definitively like to see this merged)
* `TryFrom<f64>` can now be used with `no_std`

### CODE REVIEW

**Comment by loyd** [comment]
Thanks for contributing, I've enabled the checks, fix issues, please.

File: src/lib.rs
Comment (loyd) [inline src/lib.rs]: `lib.rs` is big enough, I suggest moving this struct and new functions below to `no_std.rs` or `std_shim.rs`

File: src/lib.rs
Comment (loyd) [inline src/lib.rs]: Although it doesn't matter in the whole context, it seems there should be `abs_value >= 1f64` for `f64_significant_digits(1.0)` case.

Add tests (in new created module) that this function matches `value.abs().log10().ceil()`.

**Comment by loyd** [comment]
Some warnings are the same for the master branch, I'll fix it soon

**Comment by loyd** [comment]
Done, rebase the branch, please.

Also, can you use https://www.conventionalcommits.org/en/v1.0.0/ for commit messages?

**Comment by mversic** [comment]
> Done, rebase the branch, please.
> 
> Also, can you use https://www.conventionalcommits.org/en/v1.0.0/ for commit messages?

Ok, I'll try to do the fixes and suggestions some time today

File: src/lib.rs
Comment (mversic) [reply src/lib.rs]: done

File: src/lib.rs
Comment (mversic) [reply src/lib.rs]: added tests. I think previous implementation was faulty as well

File: src/no_std.rs
Comment (mversic) [inline src/no_std.rs:41]: I think this was incorrect previously. Please confirm

Comment (mversic) [inline src/no_std.rs:41]: I think this was incorrect previously. Please confirm

Comment (loyd) [reply src/no_std.rs:41]: Agree, the previous version can produce greater value, thanks for fixing.

File: src/no_std.rs
Comment (mversic) [inline src/no_std.rs:45]: also, we can just have this implementation for both `no_std` and `std`. It's not obvious to me that the `std` implementation is more performant

File: src/no_std.rs
Comment (mversic) [reply src/no_std.rs:41]: Could you help me make a test case in `src/tests/mod.rs::from_f64` where the conversion to fixed point results in incorrect precision? I'm not able to find a test case that covers this

File: src/no_std.rs
Comment (loyd) [reply src/no_std.rs:45]: https://godbolt.org/z/fbdd5EP96

llvm.log10.f64 generates code without loops.

Anyway, we can always write a benchmark to compare implementations)

Comment (mversic) [reply src/no_std.rs:45]: Yeah, there won't be loops with `log10` but overall complexity will remain the same since the loop runs as many times as there are digits in the string decimal representation of the number

File: src/no_std.rs
Comment (loyd) [reply src/no_std.rs:41]: I think, tests in `tests::from_f64` are enough, for now, I haven't found a counterexample in five minutes.

As an option, we can use `proptest` and check something like the following property:
```rust
for f: f64 in (-1.0, 1.0):
    let round_trip = FixedPoint::try_from(f)?.try_into::<f64>()?;
    assert!((f - round_trip).abs() < 0.1.powi(FixedPoint::PRECISION.to_i32())
```

Comment (loyd) [reply src/no_std.rs:41]: Whatever you want, it's optional, LGTM now.

But if you want to get some experience in `proptest`, I don't want to stop you)

File: src/no_std.rs
Comment (loyd) [reply src/no_std.rs:45]: > as many times as there are digits in the string decimal representation of the number

So, up to dozens of iterations at least?

Also, theoretically, can exist targets with hardware implementation, but I dunno such.

Comment (mversic) [reply src/no_std.rs:45]: Number of iterations depends on the number of digits in the integer part of the number

What I meant to say was that writing to string already has the same complexity so I don't expect this loop to degrade performance

Comment (loyd) [reply src/no_std.rs:45]: Yep, f64 → String is an expensive conversion even using `ryu`. However, I don't see reasons to use a custom unoptimized function without benches, even if this is only 5% difference.

Anywhere, the implementation with conversion sucks and definitely will be replaced with something more performant in the future.

**Comment by loyd** [comment]
Tests passed, code is good. LGTM, I'm ready to merge it.

**Review by loyd — APPROVED** [review]

File: src/no_std.rs
Comment (mversic) [reply src/no_std.rs:45]: > Anyway, we can always write a benchmark to compare implementations)

I think it would only make sense to write benchmarks in the context of the whole crate together with parsing. Comparing just these two prob won't carry as much significance.

**Comment by mversic** [comment]
@loyd I don't see merge button. Do I have the right to merge?

File: src/no_std.rs
Comment (loyd) [reply src/no_std.rs:45]: Of course, `FixedPoint::try_from` and `FixedPoint::try_into` and compare run with and without std part.

Comment (loyd) [reply src/no_std.rs:45]: @mversic, merge as is, or want to bench it?

Comment (mversic) [reply src/no_std.rs:45]: Merge as is. Maybe I bench it at some point

**Comment by loyd** [comment]
I'll update the changelog and release a new version

**Comment by loyd** [comment]
v0.6.1

**Pull request merged**
